src/main/java/frc/robot/subsystems/flywheel/Flywheel.java

```diff
@@ -0,0 +1,167 @@
+    this.io = io;
+    flywheelDisconnectedAlert = new Alert("Disconnected flywheel motor", AlertType.kError);
+  }
+
+  /**
+   * Runs the SysId Quasistatic test in the given direction for the routine specified by {@link
+   * #m_sysIdRoutineToApply}.
+   *
+   * @param direction Direction of the SysId Quasistatic test
+   * @return Command to run
+   */
+  public Command sysIdQuasistatic(SysIdRoutine.Direction direction) {
+    return m_sysIdRoutineToApply.quasistatic(direction);
+  }
+
+  /**
+   * Runs the SysId Dynamic test in the given direction for the routine specified by {@link
+   * #m_sysIdRoutineToApply}.
+   *
+   * @param direction Direction of the SysId Dynamic test
+   * @return Command to run
+   */
+  public Command sysIdDynamic(SysIdRoutine.Direction direction) {
+    return m_sysIdRoutineToApply.dynamic(direction);
+  }
+
+  @Override
+  public void periodic() {
+    io.updateInputs(inputs);
+    Logger.processInputs("Flywheel", inputs);
+    Logger.recordOutput("Flywheel/SystemState", systemState);
+
+    /* Update alerts */
+    flywheelDisconnectedAlert.set(!inputs.flywheelMotorConnected);
+
+    /* Handle statemachine logic */
+    FlywheelState nextSystemState = systemState;
+
+    if (systemState == FlywheelState.IDLE) {
+      io.setVelocity(0);
+
+      if (requestLow) {
+        nextSystemState = FlywheelState.LOW;
+      } else if (requestMedium) {
+        nextSystemState = FlywheelState.MEDIUM;
+      } else if (requestHigh) {
+        nextSystemState = FlywheelState.HIGH;
+      }
+    } else if (systemState == FlywheelState.LOW) {
+      io.setVelocity(1000);
+
+      if (!requestLow) {
+        nextSystemState = FlywheelState.IDLE;
+      }
+    } else if (systemState == FlywheelState.MEDIUM) {
+      io.setVelocity(2000);
+
+      if (!requestMedium) {
+        nextSystemState = FlywheelState.IDLE;
+      }
+    } else if (systemState == FlywheelState.HIGH) {
+      io.setVelocity(3000);
+
+      if (!requestHigh) {
+        nextSystemState = FlywheelState.IDLE;
+      }
+    }
+    if (nextSystemState != systemState) {
+      systemState = nextSystemState;
+    }
+  }
+  // Returns whether the shooter is at setpoint for the superstructure
+  public boolean atSetpoint() {
+    Logger.recordOutput("Flywheel/RPM error", desiredRPM - inputs.flywheelVelocityRpm.abs(RPM));
+    return Math.abs(desiredRPM - inputs.flywheelVelocityRpm.abs(RPM))
+        < FLYWHEEL_SETPOINT_TOLERANCE_RPM;
+  }
+
+  public void setVelocity(double velocity) {
+    io.setVelocity(velocity);
+  }
+
+  public void runCharacterization(Voltage input) {
+    io.runCharacterizationFlywheel(input);
+  }
+
+  public void requestIdle() {
+    desiredRPM = 0;
+    unsetAllRequests();
+  }
+
+  public void requestLow() {
+    desiredRPM = 2000;
+    requestLow = true;
+  }
+
+  public void requestMedium() {
+    desiredRPM = 4000;
+    requestMedium = true;
+  }
+
+  public void requestHigh() {
+    desiredRPM = 6000;
```

Comment:
these should not be here, these would be in a constants file maybe... ie there is a generic flywheel .. (most of this code) and then every specific flywheel implementation that has differenet mass, motors, etc..